File: training_data/reflection-tuning/day13_part2_2016.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

type Point struct {
	x, y int
}

func isWall(x, y, favoriteNumber int) bool {
	sum := x*x + 3*x + 2*x*y + y + y*y + favoriteNumber
	bits := 0
	for sum > 0 {
		bits += sum & 1
		sum >>= 1
	}
	return bits%2 != 0
}

func bfs(favoriteNumber, maxSteps int) (int, int) {
	start := Point{1, 1}
	target := Point{31, 39}
	queue := []Point{start}
	visited := map[Point]int{start: 0}
	directions := []Point{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if current == target {
			return visited[current], len(visited)
		}

		for _, dir := range directions {
			next := Point{current.x + dir.x, current.y + dir.y}
			if next.x < 0 || next.y < 0 || isWall(next.x, next.y, favoriteNumber) {
				continue
			}
			if _, seen := visited[next]; !seen && visited[current] < maxSteps {
				visited[next] = visited[current] + 1
				queue = append(queue, next)
			}
		}
	}

	return -1, len(visited)
}

func main() {
	file, _ := os.Open("input.txt")
	defer file.Close()
	scanner := bufio.NewScanner(file)
	scanner.Scan()
	favoriteNumber, _ := strconv.Atoi(scanner.Text())

	steps, locations := bfs(favoriteNumber, 50)
	fmt.Println("Part 1:", steps)
	fmt.Println("Part 2:", locations)
}

File: r/day13_part2_2016.R

# Function to determine if a coordinate is a wall or open space
is_wall <- function(x, y, favorite_number) {
  if (x < 0 || y < 0) return(TRUE) # Treat negative coordinates as walls
  val <- x*x + 3*x + 2*x*y + y + y*y + favorite_number
  sum(as.integer(intToBits(val))) %% 2 != 0
}

# Function to find reachable locations within a maximum number of steps
find_reachable <- function(start_x, start_y, favorite_number, max_steps) {
  visited <- matrix(FALSE, nrow = 100, ncol = 100) # Initialize a matrix to track visited locations
  queue <- list(c(start_x, start_y, 0)) # Queue of locations to explore, with steps
  visited[start_x + 1, start_y + 1] <- TRUE # Mark starting location as visited
  reachable_count <- 0
  
  while (length(queue) > 0) {
    current <- queue[[1]]
    queue <- queue[-1]
    x <- current[1]
    y <- current[2]
    steps <- current[3]
    
    if (steps <= max_steps) {
      reachable_count <- reachable_count + 1
    }
    
    if (steps < max_steps) {
      # Explore neighbors (up, down, left, right)
      neighbors <- list(c(x, y - 1), c(x, y + 1), c(x - 1, y), c(x + 1, y))
      for (neighbor in neighbors) {
        nx <- neighbor[1]
        ny <- neighbor[2]
        if (nx >= 0 && ny >= 0 && !is_wall(nx, ny, favorite_number) && !visited[nx + 1, ny + 1]) {
          queue <- append(queue, list(c(nx, ny, steps + 1)))
          visited[nx + 1, ny + 1] <- TRUE
        }
      }
    }
  }
  return(reachable_count)
}

# Read input from file
input <- readLines("input.txt")
favorite_number <- as.integer(input[1])

# Part 1: Find shortest path to 31,39
start_x <- 1
start_y <- 1
target_x <- 31
target_y <- 39

# Use BFS to find shortest path
queue <- list(c(start_x, start_y, 0))
visited <- matrix(FALSE, nrow = 100, ncol = 100)
visited[start_x + 1, start_y + 1] <- TRUE
shortest_path <- Inf

while (length(queue) > 0) {
  current <- queue[[1]]
  queue <- queue[-1]
  x <- current[1]
  y <- current[2]
  steps <- current[3]
  
  if (x == target_x && y == target_y) {
    shortest_path <- steps
    break
  }
  
  neighbors <- list(c(x, y - 1), c(x, y + 1), c(x - 1, y), c(x + 1, y))
  for (neighbor in neighbors) {
    nx <- neighbor[1]
    ny <- neighbor[2]
    if (nx >= 0 && ny >= 0 && !is_wall(nx, ny, favorite_number) && !visited[nx + 1, ny + 1]) {
      queue <- append(queue, list(c(nx, ny, steps + 1)))
      visited[nx + 1, ny + 1] <- TRUE
    }
  }
}

cat("Part 1:", shortest_path, "\n")

# Part 2: Find reachable locations within 50 steps
max_steps <- 50
reachable_count <- find_reachable(start_x, start_y, favorite_number, max_steps)
cat("Part 2:", reachable_count, "\n")
